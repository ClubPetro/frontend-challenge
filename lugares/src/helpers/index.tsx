import  capitalize  from './capitalize'

export {
    capitalize
}