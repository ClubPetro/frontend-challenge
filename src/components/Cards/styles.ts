import { makeStyles, Theme } from "@material-ui/core";

export default makeStyles((theme: Theme) => ({
  container: {
    display: "flex",
    flexWrap: "wrap",
    marginTop: "20px",
    marginBottom: "50px",
    maxWidth: "1280px",
  },
  text: {
    fontSize: "14px",
    textAlign: "center",
    marginTop: theme.spacing(1),
    [theme.breakpoints.down("xs")]: {
      marginBottom: theme.spacing(2),
    },
  },
}));
