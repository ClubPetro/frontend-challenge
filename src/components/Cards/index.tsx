import React, { useCallback, useState } from 'react'
import Dialog from '@material-ui/core/Dialog'
import DialogActions from '@material-ui/core/DialogActions'
import * as yup from 'yup'

import { TravelsProps, useTravel } from '../../hooks/travels'

import Card from '../Card'
import Input from '../Input'

import * as S from './styles'

type FormErrorProps = {
  [key: string]: boolean
}

type CardsProps = {
  travels: TravelsProps
}

const Cards = ({ travels }: CardsProps) => {
  const [local, setLocal] = useState('')
  const [target, setTarget] = useState('')
  const [open, setOpen] = useState(false)
  const [edit, setEdit] = useState(false)
  const [formErrors, setFormErrors] = useState<FormErrorProps>(
    {} as FormErrorProps
  )
  const { updateTravel } = useTravel()

  const [travelId, setTravelId] = useState(-1)
  const { removeTravel } = useTravel()

  const handleClose = useCallback(() => {
    setOpen(false)
  }, [])

  const handleClick = useCallback((id: number, type: string) => {
    setFormErrors({})
    setTarget('')
    setLocal('')
    setTravelId(id)
    setOpen(true)

    type === 'edit' ? setEdit(true) : setEdit(false)
  }, [])

  const handleConfirmModal = useCallback(
    async (id: number, type: string) => {
      switch (type) {
        case 'edit':
          try {
            const schema = yup.object().shape({
              local: yup.string().required(),
              target: yup
                .string()
                .required()
                .matches(
                  /^(1[0-2]|0[1-9]|\d)\/(20\d{2}|19\d{2}|0(?!0)\d|[1-9]\d)$/
                ),
            })

            await schema.validate(
              {
                local,
                target,
              },
              { abortEarly: false }
            )

            await updateTravel({ id: travelId, local, target })
            setOpen(false)
          } catch (err) {
            const error = err.errors.map((error: string) => {
              const field = error.split(' ')[0]

              return field
            })

            const obj = Object.fromEntries(
              error.map((error: string) => [error, true])
            )

            setFormErrors(obj)
          }
          break

        case 'remove':
          removeTravel(travelId)
          setOpen(false)
          break
      }
    },
    [local, removeTravel, travelId, updateTravel, target]
  )

  return (
    <S.Wrapper>
      {travels.length > 0 ? (
        travels.map((travel) => (
          <Card
            key={travel.id}
            flag={travel.country_flag_url}
            countryName={travel.country_name}
            local={travel.local}
            target={travel.target}
            onEdit={() => handleClick(travel.id, 'edit')}
            onRemove={() => handleClick(travel.id, 'remove')}
          />
        ))
      ) : (
        <S.NoResults>Não há resultados disponíveis :(</S.NoResults>
      )}

      {!!open && (
        <Dialog
          open={open}
          onClose={handleClose}
          aria-labelledby="alert-dialog-title"
          aria-describedby="alert-dialog-description"
        >
          <S.StyledDialogTitle>
            {'Atualize os dados do lugar'}
          </S.StyledDialogTitle>
          <S.StyledDialogContent>
            {edit ? (
              <>
                <S.StyledDialogContentText>
                  Digite os dados abaixo para atualizar o lugar que você quer
                  conhecer
                </S.StyledDialogContentText>
                <Input
                  mask=""
                  outline
                  inputSize="medium"
                  onChange={(e) => setLocal(e.target.value)}
                  value={local}
                  error={formErrors.local}
                  fullWidth
                  placeholder="Local"
                />
                <Input
                  mask="99/9999"
                  outline
                  inputSize="medium"
                  error={formErrors.target}
                  onChange={(e) => setTarget(e.target.value)}
                  value={target}
                  fullWidth
                  placeholder="mês/ano"
                />
              </>
            ) : (
              <S.StyledDialogContentText id="alert-dialog-description">
                Ao excluir esse lugar não será possível mais recuperá-lo. Deseja
                seguir com essa ação?
              </S.StyledDialogContentText>
            )}
          </S.StyledDialogContent>
          <DialogActions>
            <S.StyledButton aria-label="Close modal" onClick={handleClose}>
              Cancelar
            </S.StyledButton>
            <S.StyledButton
              aria-label="Confirm modal"
              onClick={() =>
                handleConfirmModal(travelId, edit === true ? 'edit' : 'remove')
              }
            >
              Confirmar
            </S.StyledButton>
          </DialogActions>
        </Dialog>
      )}
    </S.Wrapper>
  )
}

export default Cards
