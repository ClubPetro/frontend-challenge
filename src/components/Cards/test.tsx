import React from 'react'
import { fireEvent, screen } from '@testing-library/react'

import Cards from '.'
import { renderWithTheme } from '../../utils/tests/helpers'
import { act } from 'react-dom/test-utils'

jest.mock('../../hooks/travels', () => {
  return {
    useTravel: () => ({
      updateTravel: jest.fn(),
      removeTravel: jest.fn(),
      setMultipleTravels: jest.fn(),
    }),
  }
})

const travels = [
  {
    id: 1,
    country_name: 'Country',
    country_flag_url: '/img/country.png',
    local: 'Local',
    target: 'Target',
  },
]

describe('<Cards />', () => {
  it('should render the cards properly', async () => {
    const { container } = renderWithTheme(<Cards travels={travels} />)

    expect(screen.getByRole('img')).toHaveAttribute('src', '/img/country.png')

    expect(
      screen.getByRole('heading', { name: /country/i })
    ).toBeInTheDocument()

    expect(container.firstChild).toMatchSnapshot()
  })

  it('should open a modal if card edit button is clicked', async () => {
    renderWithTheme(<Cards travels={travels} />)

    const editButton = screen.getByLabelText(/edit card button/i)

    fireEvent.click(editButton)

    expect(
      screen.getByText(
        'Digite os dados abaixo para atualizar o lugar que você quer conhecer'
      )
    ).toBeInTheDocument()
  })

  it('should open a modal if card remove button is clicked', async () => {
    renderWithTheme(<Cards travels={travels} />)

    const removeButton = screen.getByLabelText(/remove card button/i)

    fireEvent.click(removeButton)

    expect(
      screen.getByText(
        'Ao excluir esse lugar não será possível mais recuperá-lo. Deseja seguir com essa ação?'
      )
    ).toBeInTheDocument()
  })

  it('should hide the modal if close button is clicked', async () => {
    renderWithTheme(<Cards travels={travels} />)

    const editButton = screen.getByLabelText(/edit card button/i)

    fireEvent.click(editButton)

    const localInput = screen.getByPlaceholderText(/local/i)

    const closeButton = screen.getByLabelText(/close modal/i)

    fireEvent.click(closeButton)

    expect(localInput).not.toBeInTheDocument()
  })

  it('should be able to update the card if edit modal is confirmed with valid values', async () => {
    renderWithTheme(<Cards travels={travels} />)

    const editButton = screen.getByLabelText(/edit card button/i)

    await act(async () => {
      fireEvent.click(editButton)
    })

    const localInput = screen.getByPlaceholderText(/local/i)
    const targetInput = screen.getByPlaceholderText(/mês\/ano/i)

    fireEvent.change(localInput, { target: { value: 'Local' } })
    fireEvent.change(targetInput, { target: { value: '01/2021' } })

    const confirmButton = screen.getByLabelText(/confirm modal/i)

    await act(async () => {
      fireEvent.click(confirmButton)
    })

    expect(targetInput).not.toBeInTheDocument()
  })

  it('should NOT be able to update the card if edit modal is confirmed with invalid values', async () => {
    renderWithTheme(<Cards travels={travels} />)

    const editButton = screen.getByLabelText(/edit card button/i)

    fireEvent.click(editButton)

    const confirmButton = screen.getByLabelText(/confirm modal/i)

    await act(async () => {
      fireEvent.click(confirmButton)
    })

    expect(screen.getByPlaceholderText(/local/i).parentElement).toHaveStyleRule(
      'border-bottom',
      '0.3rem solid red',
      {
        modifier: '::before',
      }
    )
  })

  it('should be able to remove the selected card', async () => {
    renderWithTheme(<Cards travels={travels} />)

    const removeButton = screen.getByLabelText(/remove card button/i)

    fireEvent.click(removeButton)

    const confirmButton = screen.getByLabelText(/confirm modal/i)

    await act(async () => {
      fireEvent.click(confirmButton)
    })

    expect(confirmButton).not.toBeInTheDocument()
  })

  it('should render a message if there is no travels available', async () => {
    renderWithTheme(<Cards travels={[]} />)

    expect(
      screen.getByText(/não há resultados disponíveis/i)
    ).toBeInTheDocument()
  })
})
