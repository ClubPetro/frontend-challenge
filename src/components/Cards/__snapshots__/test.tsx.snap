// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Cards /> should render the cards properly 1`] = `
.c1 {
  display: grid;
  grid-template-rows: 0.6fr auto 1fr;
  height: 25rem;
  width: 25rem;
  padding: 1.6rem;
  border-radius: 1rem;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c12 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  padding: 0 0.8rem;
}

.c12 p + p {
  margin-top: calc(0.8rem + 0.4rem);
}

.c11 {
  border: 0;
  border-top: 1px solid #ABABAB;
  margin-top: 0.8rem;
}

.c6 svg + svg {
  margin-left: 1.6rem;
}

.c4 {
  max-width: 80%;
}

.c4 .c9 {
  margin-top: 1.6rem;
}

.c5 {
  width: 5.6rem;
  height: 3.4rem;
}

.c10 {
  font-size: 1.6rem;
  text-transform: uppercase;
  font-weight: 700;
  color: #4F9419;
}

.c7 {
  font-size: 1.8rem;
  color: #868686;
  cursor: pointer;
}

.c8 {
  font-size: 1.8rem;
  color: #868686;
  cursor: pointer;
}

.c0 {
  display: grid;
  grid-template-columns: repeat(auto-fit,minmax(250px,1fr));
  grid-gap: 3.2rem;
  width: 100%;
  justify-items: center;
  padding: 5.6rem 4.0rem;
  max-width: 150rem;
  margin: 0 auto;
}

@media (min-width:1440px) {
  .c0 {
    grid-template-columns: repeat(5,1fr);
    justify-items: center;
  }
}

<div
  class="c0"
>
  <div
    class="MuiPaper-root MuiCard-root c1 MuiPaper-elevation4 MuiPaper-rounded"
  >
    <div
      class="c2"
    >
      <div
        class="c3"
      >
        <div
          class="c4"
        >
          <img
            alt="Country"
            class="c5"
            src="/img/country.png"
          />
        </div>
        <div
          class="c6"
        >
          <svg
            aria-hidden="true"
            aria-label="Edit card button"
            class="MuiSvgIcon-root c7"
            focusable="false"
            role="button"
            viewBox="0 0 24 24"
          >
            <path
              d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.9959.9959 0 00-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
            />
          </svg>
          <svg
            aria-hidden="true"
            aria-label="Remove card button"
            class="MuiSvgIcon-root c8"
            focusable="false"
            role="button"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
            />
          </svg>
        </div>
      </div>
      <h3
        class="c9 c10"
      >
        Country
      </h3>
    </div>
    <hr
      class="c11"
    />
    <div
      class="c12"
    >
      <p>
        Local: 
        Local
         
      </p>
      <p>
        Meta: 
        Target
         
      </p>
    </div>
  </div>
</div>
`;
